Examples/Tests/particle_fields_diags/analysis_particle_diags_impl.py

```diff
@@ -0,0 +1,162 @@
+        uzavg[x_ind[i_p],y_ind[i_p],z_ind[i_p]] += uz[i_p] * w[i_p]
+        zuzavg[x_ind[i_p],y_ind[i_p],z_ind[i_p]] += z[i_p] * uz[i_p] * w[i_p]
+        wavg[x_ind[i_p],y_ind[i_p],z_ind[i_p]] += w[i_p]
+
+    wavg_adj = np.where(wavg == 0, 1, wavg)
+    values_yt['protons: zavg'] = zavg / wavg_adj
+    values_yt['protons: uzavg'] = uzavg / wavg_adj
+    values_yt['protons: zuzavg'] = zuzavg / wavg_adj
+
+    # Photons (momentum in units of m_e c)
+    x = ad['photons', 'particle_position_x'].to_ndarray()
+    y = ad['photons', 'particle_position_y'].to_ndarray()
+    z = ad['photons', 'particle_position_z'].to_ndarray()
+    uz = ad['photons', 'particle_momentum_z'].to_ndarray() / m_e / c
+    w  = ad['photons', 'particle_weight'].to_ndarray()
+
+    x_ind = ((x - ds.domain_left_edge[0].value) / dx[0]).astype(int)
+    y_ind = ((y - ds.domain_left_edge[1].value) / dx[1]).astype(int)
+    z_ind = ((z - ds.domain_left_edge[2].value) / dx[2]).astype(int)
+
+    zavg = np.zeros(ds.domain_dimensions)
+    uzavg = np.zeros(ds.domain_dimensions)
+    zuzavg = np.zeros(ds.domain_dimensions)
+    wavg = np.zeros(ds.domain_dimensions)
+
+    for i_p in range(len(x)):
+        zavg[x_ind[i_p],y_ind[i_p],z_ind[i_p]] += z[i_p] * w[i_p]
+        uzavg[x_ind[i_p],y_ind[i_p],z_ind[i_p]] += uz[i_p] * w[i_p]
+        zuzavg[x_ind[i_p],y_ind[i_p],z_ind[i_p]] += z[i_p] * uz[i_p] * w[i_p]
+        wavg[x_ind[i_p],y_ind[i_p],z_ind[i_p]] += w[i_p]
+
+    wavg_adj = np.where(wavg == 0, 1, wavg)
+    values_yt['photons: zavg'] = zavg / wavg_adj
+    values_yt['photons: uzavg'] = uzavg / wavg_adj
+    values_yt['photons: zuzavg'] = zuzavg / wavg_adj
+
+
+    values_rd = dict()
+    # Load reduced particle diagnostic data from plotfiles
+    values_rd['electrons: zavg'] = ad0[('boxlib','z_electrons')]
+    values_rd['protons: zavg'] = ad0[('boxlib','z_protons')]
+    values_rd['photons: zavg'] = ad0[('boxlib','z_photons')]
+
+    values_rd['electrons: uzavg'] = ad0[('boxlib','uz_electrons')]
+    values_rd['protons: uzavg'] = ad0[('boxlib','uz_protons')]
+    values_rd['photons: uzavg'] = ad0[('boxlib','uz_photons')]
+
+    values_rd['electrons: zuzavg'] = ad0[('boxlib','zuz_electrons')]
+    values_rd['protons: zuzavg'] = ad0[('boxlib','zuz_protons')]
+    values_rd['photons: zuzavg'] = ad0[('boxlib','zuz_photons')]
+
+    #--------------------------------------------------------------------------------------------------
+    # Part 3: compare values from plotfiles and diagnostics and print output
+    #--------------------------------------------------------------------------------------------------
+
+    error = dict()
+    tolerance = 5e-3 if single_precision else 1e-12
+
+    # The comparison of field energies requires a large tolerance,
+    # because the field energy from the plotfiles is computed from cell-centered data,
+    # while the field energy from the reduced diagnostics is computed from (Yee) staggered data.
```

Comment:
This comment about field energies might be a residual of the original Python analysis script, right? If it's not relevant for the particle properties tested here, we could probably remove it.